### PR TITLE
[ADT][NFC] Refactor/optimize DenseMap::copyFrom

### PR DESCRIPTION
**Pull request by marcauberer**
Simplify code, constexpr-ify if with constexpr condition

### CODE REVIEW

File: llvm/include/llvm/ADT/DenseMap.h [hunk 1]
```diff
@@ -471,19 +471,22 @@ class DenseMapBase : public DebugEpochBase {
...
-        ::new (&getBuckets()[i].getFirst())
-            KeyT(other.getBuckets()[i].getFirst());
-        if (!KeyInfoT::isEqual(getBuckets()[i].getFirst(), getEmptyKey()) &&
-            !KeyInfoT::isEqual(getBuckets()[i].getFirst(), getTombstoneKey()))
-          ::new (&getBuckets()[i].getSecond())
-              ValueT(other.getBuckets()[i].getSecond());
+    } else {
+      const KeyT EmptyKey = getEmptyKey();
+      const KeyT TombstoneKey = getTombstoneKey();
+      for (size_t I = 0; I < getNumBuckets(); ++I) {
```
Comment (kuhar) [inline llvm/include/llvm/ADT/DenseMap.h:483]: If the trip count is constant here, this should only calculate it once. See https://llvm.org/docs/CodingStandards.html#don-t-evaluate-end-every-time-through-a-loop.

Comment (marcauberer) [reply llvm/include/llvm/ADT/DenseMap.h]: Good point, thanks.

File: llvm/include/llvm/ADT/DenseMap.h [hunk 2]
```diff
@@ -471,19 +471,23 @@ class DenseMapBase : public DebugEpochBase {
...
-        ::new (&getBuckets()[i].getFirst())
-            KeyT(other.getBuckets()[i].getFirst());
-        if (!KeyInfoT::isEqual(getBuckets()[i].getFirst(), getEmptyKey()) &&
-            !KeyInfoT::isEqual(getBuckets()[i].getFirst(), getTombstoneKey()))
-          ::new (&getBuckets()[i].getSecond())
-              ValueT(other.getBuckets()[i].getSecond());
+    } else {
+      const KeyT EmptyKey = getEmptyKey();
+      const KeyT TombstoneKey = getTombstoneKey();
+      const size_t NumBuckets = getNumBuckets();
```
Comment (kazutakahirata) [inline llvm/include/llvm/ADT/DenseMap.h:483]: nit: If you are placing `getBuckets()` just before the if-then-else block, you might move `getNumBuckets()` there as well for consistency.  If you are moving it, don't forget to update the `memcpy` above also.

Comment (marcauberer) [reply llvm/include/llvm/ADT/DenseMap.h]: Done.

**Review by kazutakahirata — APPROVED** [review]
LGTM.  Thanks!

**Pull request merged**